### ISSUE TITLE
Metadata discovery

### ISSUE THREAD

**Issue by kyessenov** [istio/istio#42070]
Part of https://github.com/istio/istio/issues/41527.

Proxy receives metadata out-of-band instead of on the wire. This has two major advantages:
* interop between any mTLS with telemetry
* interop with xDS planes, without custom LB programming
* improved latency/bandwidth, with less bytes on the network

Implementation:
* re-use ambient workload discovery in on-demand delta-mode.
* async telemetry reporting in native stats filter, delay metrics push until metadata is in the proxy cache.
* this will not be implemented for Wasm filters, due to limitations.
* implement generic async logging in envoy https://github.com/envoyproxy/envoy/issues/10326, for OTel logger.
* disable MX filters in istio.
* may restrict to TCP first (since it pushes at 5s, no additional telemetry latency added).

cc @aryan16  @zirain

**Comment by howardjohn** [comment]
I think we will need some negotiation (ALPN?) to disable MX for backwards compat with old proxies. Or keep it around for a few releases first

**Comment by kyessenov** [comment]
Agree. We don't really support long living old versions of Envoy, apart from the upgrade process (which can force restart the application workloads), so one release is sufficient I think.

**Comment by howardjohn** [comment]
That might be the case but its not something we have solidified before; we have not broken data plane version skew support for many years. We regularly have folks upgrading from massive version skews like 1.4 to 1.16 with data plane connectivity.

So - it may be reason to say 1 release (or some X), but if we do we should have a broader discussion since it will be the first time we make that decision. At the version least we official support skipping one version, so it would need to be 2 versions at minimum.

**Comment by kyessenov** [comment]
Well if you have to support direct connectivity with Istio 1.4 then clearly the data path can never evolve, so that would be a very unreasonable policy decision :) This would make sense if Istio used standard protocols, but the ALPN/TLS hacks are not used anywhere else, so that bar is not met. There's no problem keeping old MX code around, and the control plane can just program both at the same time.

**Comment by howardjohn** [comment]
I am not saying that _is_ the policy, I am saying there is none, so if we want to be the first to break it we will also need to come up with the policy and get approval from a broader group

**Comment by howardjohn** [comment]
We have never broken compatibility. This is proposing to break it. We need TOC approval to do so.

**Comment by kyessenov** [comment]
Compatibility with what? We're not changing any Istio API (and the only one affected which is Telemetry is not default anyways), this is just an internal mechanism.

**Comment by howardjohn** [comment]
Communication between different proxy version. Its not just an internal mechanism since this will have user facing impact on users running Istio `v1.old` with Istio `v1.new`.

**Comment by kyessenov** [comment]
The communication interface is regular mTLS without MX ALPN token. ALPN specification requires fallback so istio must fallback to the less capable protocol (regular mTLS). gRPC and others interoperate this way since they don't have MX network filter.

**Comment by howardjohn** [comment]
We need (if we want full compat) the old clients to actually do MX though, not just fallback, or their telemetry will be broken. We can chose not to do that (and thus change user facing behavior), if we get TOC approval.

**Comment by kyessenov** [comment]
Breaking client telemetry never stopped us from declaring protocol support (e.g. it doesn't work for gRPC or HBONE), so it would be inconsistent to hold that as a gate if it wasn't even used as a gate to add a feature. Telemetry was always add-on and never required for protocol support, which is why making changes to the telemetry pipeline is not a compatibility violation. The important telemetry (audit logging) is always server-side and that doesn't use MX.

**Comment by howardjohn** [comment]
I appreciate your perspective, but  I am asking you to please have the TOC review the design doc. Its okay if we don't agree on whether that review is needed or not, but the implementation cannot be merged without it.

**Comment by kyessenov** [comment]
I agree that TOC needs to review the item: "remove MX support", which I'm sure can be delayed after all Envoy work is done. That's just the icing on the cake, I think the native capability of Envoy to fetch metadata has a greater value outside of this Istio feature.

**Comment by kyessenov** [comment]
A few additional requirements:
* Recommend this for small/default installations as well as installations using telemetry collectors.
* Support graceful degradation with "workload key" as the value;
* Define a principled guidance for the "split-brain" control planes (multi-network or multi-cluster setups with disjoint service registries)
* Support workload trustworthy workload keys (pod IDs)
